src/protocol/sort/generate_permutation.rs

```diff
@@ -259,28 +412,95 @@ mod tests {
         let h1_future =
-            shuffle_and_reveal_permutation(ctx1.narrow("shuffle_reveal"), BATCHSIZE, perm1);
+            shuffle_and_reveal_permutation(ctx1.narrow("shuffle_reveal"), BATCHSIZE, perm1, None);
         let h2_future =
-            shuffle_and_reveal_permutation(ctx2.narrow("shuffle_reveal"), BATCHSIZE, perm2);
+            shuffle_and_reveal_permutation(ctx2.narrow("shuffle_reveal"), BATCHSIZE, perm2, None);
 
         let perms_and_randoms = join3(h0_future, h1_future, h2_future).await;
 
-        assert_eq!(perms_and_randoms[0].revealed, perms_and_randoms[1].revealed);
-        assert_eq!(perms_and_randoms[1].revealed, perms_and_randoms[2].revealed);
+        assert_eq!(
+            perms_and_randoms[0].reveal_and_random_permutation.revealed,
+            perms_and_randoms[1].reveal_and_random_permutation.revealed
+        );
+        assert_eq!(
+            perms_and_randoms[1].reveal_and_random_permutation.revealed,
+            perms_and_randoms[2].reveal_and_random_permutation.revealed
+        );
 
         assert_eq!(
-            perms_and_randoms[0].randoms_for_shuffle.0,
-            perms_and_randoms[2].randoms_for_shuffle.1
+            perms_and_randoms[0]
+                .reveal_and_random_permutation
+                .randoms_for_shuffle
+                .0,
+            perms_and_randoms[2]
+                .reveal_and_random_permutation
+                .randoms_for_shuffle
+                .1
         );
         assert_eq!(
-            perms_and_randoms[1].randoms_for_shuffle.0,
-            perms_and_randoms[0].randoms_for_shuffle.1
+            perms_and_randoms[1]
+                .reveal_and_random_permutation
+                .randoms_for_shuffle
+                .0,
+            perms_and_randoms[0]
+                .reveal_and_random_permutation
+                .randoms_for_shuffle
+                .1
         );
         assert_eq!(
-            perms_and_randoms[2].randoms_for_shuffle.0,
-            perms_and_randoms[1].randoms_for_shuffle.1
+            perms_and_randoms[2]
+                .reveal_and_random_permutation
+                .randoms_for_shuffle
+                .0,
+            perms_and_randoms[1]
+                .reveal_and_random_permutation
+                .randoms_for_shuffle
+                .1
         );
     }
+
+    #[tokio::test]
+    pub async fn malicious_sort_in_semi_honest() {
+        const COUNT: usize = 5;
+
+        let world = TestWorld::new(QueryId);
+        let mut rng = thread_rng();
+
+        let mut match_keys = Vec::with_capacity(COUNT);
+        match_keys.resize_with(COUNT, || MaskedMatchKey::mask(rng.gen()));
+
+        let mut expected = match_keys
+            .iter()
+            .map(|mk| u64::from(*mk))
+            .collect::<Vec<_>>();
+        expected.sort_unstable();
+
+        let result = world
+            .semi_honest(
+                match_keys.clone(),
+                |ctx: SemiHonestContext<Fp31>, mk_shares| async move {
```

Comment:
Do you really need the type specification here?

Reply:
It needs the type for field and not the context actually.